File: java/LeetCode_Problems/easy/1_twoSum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        return optimal(nums, target); 
        /*
        for (int i = 0; i < nums.length; i++) {
            for (int j = 0; j < nums.length; j++) {
                if (nums[i] + nums[j] == target && i != j) {
                    return new int[] {i, j};
                }
            }
        }
        return new int[] {0, 0};
        */
    }
    
    public int[] optimal (int[] nums, int target) {
        //complement, position
        Map<Integer, Integer> map = new HashMap<>(); 
        
        for (int i = 0; i < nums.length; i++) {
            map.put(nums[i], i);
        }
        
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            
            if (map.containsKey(complement) && i != map.get(complement)) {
                return new int[] {i, map.get(complement)};
            }
        }
        return null; 
    }
}

File: python/leetcode_problems/easy/1_two_sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        if (len(nums) == 0): 
            return []

        dict = {}
        for idx, i in enumerate(nums): 
            complement = target - i
            dict[complement] = idx
        
        for idx, i in enumerate(nums): 
            if (i in dict and dict[i] != idx):
                return [idx, dict[i]]
        return [0,0]

    def twoSum_poorly_written(self, nums: List[int], target: int) -> List[int]:
        if (len(nums) == 0): 
            return []
        
        """
        Best solution: 
            We want to iterate over a list and effectively find the complement solution. 

            1. Dictionary: iterate over list and set the complement values
            {complement, index}
            2. Iterate over the list again and if the complement exists in the dictionary, return the value of the key and the current key. (note: current index and complement index cannot be the same)
        """
        idx = 0
        dict = {}
        for i in nums: 
            complement = target - i
            dict[complement] = idx
            idx = idx + 1
        
        idx = 0
        for i in nums: 
            if (i in dict):
                if (dict[i] != idx):
                    return [dict[i], idx]
            idx = idx + 1

        return [1,2]
